src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -960,6 +952,23 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                                 value, overloaded);
     }
 
+    template <typename T>
+    T get_constant_value(ASR::expr_t *expr, const Location &loc) {
+        ASR::ttype_t *type = ASRUtils::expr_type(expr);
+        if (ASRUtils::is_integer(*type)) {
+            return ASR::down_cast<ASR::IntegerConstant_t>(ASRUtils::expr_value(expr))->m_n;
+        }
+        else if (ASRUtils::is_real(*type)) {
+            return ASR::down_cast<ASR::RealConstant_t>(ASRUtils::expr_value(expr))->m_r;
+        }
+        else if (ASRUtils::is_logical(*type)) {
+            return ASR::down_cast<ASR::LogicalConstant_t>(ASRUtils::expr_value(expr))->m_value;
+        }
+        else {
+            throw SemanticError("Cannot get constant value of " + ASRUtils::type_to_str_python(type), loc);
+        }
```

Comment:
This is slower however, as previously it simply down casted, which is a free operation in Release mode, while now you have to do a linear search for the right type using the if statements. So we don't want to do that.

However, there might be a way to achieve the same speed by somehow using templates and casting right away without the linear search.

Reply:
May be we can use enumeration and then switch operator to avoid searching.

Reply:
Thanks for pointing!  @certik @Smit-create 

> May be we can use enumeration and then switch operator to avoid searching.

Let me try this.

Reply:
Even switch is slower. Compilers often do a jump table instead of a linear search for switches, but it is still slower than the essentially free downcast. 

The way to do this is with templates --- you can specialize the templated function for `double` and in the specialization cast to `RealConstant_t` directly. Then it should be as fast. It does seems like quite a lot of complication to save a few characters, so you might as well just create a function like `get_integer_value` for all types at that point. Or just leave the original code as is. :)

Reply:
No worries! I am very happy that you are thinking about how to simplify the code. We definitely want to extract as much stuff as we can into libasr, and share with LFortran (and other future frontends down the road), which will greatly simplify the number of unique code we need to maintain as part of LPython. We just have to be careful about performance.